explorer/interpreter/value.cpp

```diff
@@ -38,6 +38,19 @@ auto StructValue::FindField(std::string_view name) const
   return std::nullopt;
 }
 
+NominalClassValue::NominalClassValue(
+    Nonnull<const Value*> type, Nonnull<const Value*> inits,
+    std::optional<Nonnull<const NominalClassValue*>> base,
+    const VTable** const vptr)
+    : Value(Kind::NominalClassValue),
+      type_(type),
+      inits_(inits),
+      base_(base),
+      vptr_(vptr) {
+  // Update object's vptr to match the type's vtable.
```

Comment:
IIUC, this is updating the vptr on the pointer object to point at the type's vtable.

However, I don't understand why this approach is chosen. Why store a vptr at all, versus using the vtable of the stored type? In the explorer, I think a similar purpose is served.

If kept, I think the mutation of vptr needs to be documented in the header -- the mutation isn't something I'd expect from reading the API call alone.

Reply:
For context, I initially use a simple VTable*. It appeared like it was working (in case you saw it), but I'm pretty sure it was a side effect of an issue with `BaseElement` I had at the time.
Here:
* For a given object, each class in the hierarchy is represented by its own `NominalClassValue`
* The subtyping resolves the type queried by returning the corresponding `NominalClassValue`, and we don't have access there to child class value, so each value needs a vptr (or a single one in the parent-most class value)
* NominalClassValue construction is done recursively in `ConvertStructToClass`, which builds each level, and return a `const`. Same when setting a field in `SetFieldImpl`. Changing it to return non-const values would impact a _lot_ of code (though maybe OK?)
So my only option was to have a `VTable**` which works around the immutability of those layers by allowing to properly update the vptr as we build. Ideally, we would know from the start, but the recursive build steps we have don't allow passing additional `args` or anything AFAIK. Agree about documenting it at a minimum. Does that make sense?

Reply:
Commented on this separately, I think I may have a smaller suggestion that'd improve it for me.